src/backend/dx12/src/command.rs

```diff
@@ -1283,61 +1285,82 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
+                r.image_extent.height
+            } else {
+                r.buffer_height
+            };
 
-                assert!(buffer_width >= width as u32);
-                assert_eq!(region.buffer_offset % d3d12::D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT as u64, 0);
+            if aa != image::AaMode::Single ||
+                r.image_layers.aspects.contains(Aspects::DEPTH) ||
+                r.image_layers.aspects.contains(Aspects::STENCIL)
+            {
+                if r.image_offset.x + r.image_offset.y + r.image_offset.z != 0 ||
+                    r.image_extent.width != width as u32 || r.image_extent.height != height as u32 || r.image_extent.depth != depth as u32
+                {
+                    error!("MSAA and depth/stencil copies have to cover the whole subresource");
+                }
+            }
 
-                let row_pitch = div(buffer_width, image.block_dim.0 as _) * image.bytes_per_block as u32;
-                let slice_pitch = div(buffer_height, image.block_dim.1 as _) * row_pitch;
-                assert_eq!(row_pitch % d3d12::D3D12_TEXTURE_DATA_PITCH_ALIGNMENT as u32, 0);
+            assert!(buffer_width >= width as u32);
+            let row_pitch = div(buffer_width, image.block_dim.0 as _) * image.bytes_per_block as u32;
+            let slice_pitch = div(buffer_height, image.block_dim.1 as _) * row_pitch;
+            assert_eq!(row_pitch % d3d12::D3D12_TEXTURE_DATA_PITCH_ALIGNMENT as u32, 0);
 
-                let height = height as _;
-                let depth = depth as _;
+            let width_offset_bytes = div(r.image_offset.x as _, image.block_dim.0 as _) * image.bytes_per_block as u32;
+            assert!(width_offset_bytes <= r.buffer_offset as u32);
 
+            let (footprint_height, height_offset) = if r.image_extent.depth <= 1 {
+                // when a single depth slice is copied, we can pretend to deal with a smaller image
+                (r.image_extent.height, 0)
+            } else {
+                (height as u32, r.image_offset.y as u32)
+            };
+
+            let src_box = d3d12::D3D12_BOX {
+                left: r.image_offset.x as _,
+                top: height_offset as _,
+                right: (r.image_offset.x as u32 + r.image_extent.width) as _,
+                bottom: (height_offset + r.image_extent.height) as _,
+                front: 0,
+                back: r.image_extent.depth as _,
+            };
+
+            // Copy each layer in the region
+            for layer in r.image_layers.layers.clone() {
                 // Advance buffer offset with each layer
                 *unsafe { src.u.PlacedFootprint_mut() } = d3d12::D3D12_PLACED_SUBRESOURCE_FOOTPRINT {
-                    Offset: region.buffer_offset as UINT64 + (layer as u32 * slice_pitch * depth) as UINT64,
+                    Offset: r.buffer_offset as UINT64 -
+                        (width_offset_bytes + height_offset * row_pitch) as UINT64 +
```

Comment:
Looks suspicious, I don't believe that the resulting value will be aligned correctly